tests/catalog/test_hive.py

```diff
@@ -277,7 +277,7 @@ def test_create_table(table_schema_simple: Schema, hive_database: HiveDatabase,
             )
         ],
         current_schema_id=0,
-        last_partition_id=1000,
+        last_partition_id=999,
```

Comment:
See https://github.com/apache/iceberg-python/pull/245/files#r1479265846

Reply:
I think this one is a bit odd since it is V2 table. I had to dig into the code myself a bit as well. I noticed that the `last_partition_id` is optional in the metadata. What do you think of the following solution: https://github.com/amogh-jahagirdar/iceberg-python/pull/1

Reply:
I checked V2 unpartitioned table created by spark-iceberg-runtime and the `last_partition_id` stored in the metadata is `999`. 
<img width="494" alt="Screenshot 2024-02-21 at 16 27 58" src="https://github.com/apache/iceberg-python/assets/140284484/67f38679-5aa1-4ce3-8f43-7e87e941f09d">
Therefore I suggested to update the `last_partition_id()` in pyiceberg to align with the java implementation. 

In general, I think `999` is spec compliant since it is for `UnpartitionedSpec`, where there is no existing partition field. It implies that 1000 will be the id for the first valid partition field and thus align with the spec. Do these sound reasonable? Appreciate your thoughts on this!

Reply:
@Fokko I took a look and integrated the changes, and after going back/forth I think I'd like to keep the changes as is.
My rationale is that even after those changes there's some more workarounds that need to be done to make sure new IDs start at 1000 (after taking the changes directly, field IDs for non-REST will start at 1001 without any more changes).

Now, technically it does not seem to be a hard requirement that ids need to start at 1000 for v2 tables. Even for v1 starting at 1000 does not seem to be a requirement, rather just how the Java lib was implemented.

```
In v1, partition field IDs were not tracked, but were assigned sequentially starting at 1000 in the reference implementation.
```
I read this as "this was how we originally implemented in Java but not really required so long as it's unique and increasing for new fields"

All that said, I'd advocate for just following the practice of starting at 1000 for both v1 and v2.

On returning 999 for unpartitioned spec:

As @HonahX alluded to I think that makes for a logical API, considering we want to start a field at 1000, the unpartitioned spec (no fields) last assigned field ID should be one less than that. I don't think we want to return 1000 in that case. This is also what Spark sets for the unpartitioned spec, and is spec compliant (since the spec doesn't mandate any particular IDs)

I could see the argument where we just want to return None for a last_assigned_partition_id for this API but that just shifts the responsibility to other locations to make sure Ids are set correctly according to our scheme which ends up being more messy imo compared to just defining the API to return 999 if the only spec is the unpartitioned spec (which is the value which would be set anyways).

I also think it makes sense to set last-assigned-partition-id for both V1 and V2. Even though it's optional for V1 we set the other optional fields for v1 metadata so it seems a bit odd to make this particular case an exception.

Reply:
Sounds reasonable to me, and as long as it is in line with the spec, I'm okay with it 👍